linux/ibus-keyman/src/keymanutil.c

```diff
@@ -212,6 +212,30 @@ get_engine_for_language(
   return engine_desc;
 }
 
+int
+keyman_compare_version(const gchar *version1str, const gchar *version2str) {
+  size_t len1 = strlen(version1str);
+  size_t len2 = strlen(version2str);
+  for (size_t i = 0, j = 0; i < len1 || j < len2; i++, j++) {
+      int version1 = 0, version2 = 0;
+
+      while (i < len1 && version1str[i] != '.') {
+        version1 = version1 * 10 + (version1str[i] - '0');
+        i++;
+      }
+      while (j < len2 && version2str[j] != '.') {
+        version2 = version2 * 10 + (version2str[j] - '0');
+        j++;
+      }
+
+      if (version1 < version2)
+        return -1;
+      if (version1 > version2)
+        return +1;
+    }
+  return 0;
+}
```

Comment:
This looks okay. But it might be cleaner to do it with pointers:

```suggestion
int
keyman_compare_version(const gchar *ver1, const gchar *ver2) {
  for (; *ver1 || *ver2; ) {
    int version1 = 0, version2 = 0;

    while (*ver1 && *ver1 != '.') {
      assert(*ver1 >= '0' && *ver1 <= '9');
      version1 = version1 * 10 + (*ver1 - '0');
      ver1++;
    }
    while (*ver2 && *ver2 != '.') {
      assert(*ver2 >= '0' && *ver2 <= '9');
      version2 = version2 * 10 + (*ver2 - '0');
      ver2++;
    }

    if (version1 < version2)
      return -1;
    if (version1 > version2)
      return +1;
    if(*ver1) 
      ver1++;
    if(*ver2)
      ver2++;
  }
  return 0;
}
```

Reply:
Done. Also added a few more tests.